File: observer/weather-station/cpp/observers.cpp
#include <iostream>

#include "observers.h"
#include "stations.h"

CurrentConditions::CurrentConditions(Observable* obs) {
	observable = obs;
	observable->addObserver(this);
	temperature = humity = pressure = 0;
}

void CurrentConditions::update(Observable* observable) {
	WeatherData* wd = (WeatherData*) observable;
	temperature = wd->getTemperature();
	humity = wd->getHumidy();
	pressure = wd->getPressure();
	display();
}

void CurrentConditions::display() {
	std::cout << "\nCurrent Conditions\n";
	std::cout << temperature << "°C" << "\n";
	std::cout << humity << "°%" << "\n";
	std::cout << pressure << " mbar" << "\n";
	std::cout << std::endl;
}

StaticticsDisplay::StaticticsDisplay(Observable* observable) {
	observable->addObserver(this);
	this->observable = observable;
	temperature = humity = pressure = 0;
}

void StaticticsDisplay::update(Observable* observable) {
	
	WeatherData* wd = (WeatherData*) observable;
	temperature = wd->getTemperature();
	humity = wd->getHumidy();
	pressure = wd->getPressure();
	display();
}

void StaticticsDisplay::display() {
	std::cout << "\nStatictics Display\n";
	std::cout << temperature << "°C" << "\n";
	std::cout << humity << "°%" << "\n";
	std::cout << pressure << " mbar" << "\n";
	std::cout << std::endl;
}

File: observer/weather-station/python/observers.py
from interfaces import Observer, DisplayElement


class CurrentConditions(Observer, DisplayElement):

	def __init__(self, observable):
		self.observable = observable
		self.observable.addObserver(self)
		self.temperature = self.humity = self.pressure = 0

	def update(self, observable, *args, **kwargs):
		self.temperature = observable.getTemperature()
		self.humity = observable.getHumidy()
		self.pressure = observable.getPressure()
		self.display()

	def display(self):
		print("Current Conditions")
		print(self.temperature, self.humity, self.pressure)
		print()


class StaticticsDisplay(Observer, DisplayElement):

	def __init__(self, observable):
		self.observable = observable
		self.observable.addObserver(self)
		self.temperature = self.humity = self.pressure = 0

	def update(self, observable, *args, **kwargs):
		self.temperature = observable.getTemperature()
		self.humity = observable.getHumidy()
		self.pressure = observable.getPressure()
		self.display()

	def display(self):
		print("\nStatictics Display")
		print(self.temperature, self.humity, self.pressure)
		print()
